### PR TITLE
Task/CSXL-597: Recurring Office Hours Events

### LINKED ISSUE #597: Add Recurring Office Hours Events
## Synopsis

The new office hours feature enables instructors and TAs to create office hours events for their classes! The problem, is, you cannot make recurring events. Dealing with recurring events correctly is a large challenge, since events are linked to each other.

***Your mission, should you choose to accept it, is to implement the ability to add recurring office hours events for the XL site.***

## Acceptance Criteria

- [ ] Creating a recurring event should give a start and end date, as well as options to select days of the week.

- [ ] Deleting a recurring event should give a dialog asking whether or not to delete all future recurring events. Pressing "yes" should delete the selected event and all future ones. Pressing "no" should delete only one event. Note that the events following the selected event should still be linked such that deleting another event should give the same options and still work.

- [ ] Editing a recurring event should give a dialog asking whether or not to update all future recurring events. Pressing "yes" should update the selected event and all future ones. Pressing "no" should update only the selected event. Then, the selected event should no longer be considered a recurring event. Consider this chain of recurring events:
  ```
  [ 1 ] --> [ 2 ] --> [ 3 ] --> [ 4 ] --> [ 5 ]
  ```
  Updating `[ 2 ]` should ensure that event is no longer part of the chain, but that `[ 3 ]` is considered a recurring event of `[ 1 ]`. Therefore, choosing to delete `[ 1 ]` should give the option to delete `[ 3 ]`- `[ 5 ]` as well.

- [ ] When updating the days of the week of a recurring event and selecting "yes" to update future recurrences, these changes should be propagated, creating and deleting future events as necessary. For example, if OH are scheduled MWF and a Monday event is updated to TTh, the W and F events should be deleted, and T Th added.

## First Steps

If your team chooses to accept this, please list yourself as "assignees" to this ticket! Then, begin to work on a backend entity diagram design and a frontend design on Figma. Then, comment them here on this ticket for approval! From there, you can begin creating a branch for this feature.

Work on a branch called `reccuring-oh-events`.

For approvals, we will work in this order:
1. **Backend Review:** Tests to ensure your backend is working properly.
2. **Frontend Review:** Ensures your frontend looks good!

For tips on what we will look for, check out the [Contributing Guidelines](https://github.com/unc-csxl/csxl.unc.edu/blob/main/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by jadekeegan**
This PR adds functionality for recurring OH events on the backend and frontend. Specifically, it allows for the creation and deletion of recurring events.

## Testing
1. Click on the button to create a new event for a course site.
2. Set the recurring events option to "weekly".
3. Select which days of the week the event should recur on.
4. Select an end date for the recurrence.
5. Ensure that all expected events are created (either in the current/future section).

## Future Development
- When deleting a recurring event, create a UI to show the next few events in the pattern that will be deleted.
- Add update functionality for recurring events.
- Restrict end date to end of term / two weeks after EOT.

> _This PR resolves #597._

### CODE REVIEW

File: frontend/src/app/my-courses/course/office-hours/office-hours-editor/office-hours-editor.component.ts [hunk 1]
```diff
@@ -59,15 +56,23 @@ export class OfficeHoursEditorComponent {
   /* Holds the virtual room. */
   virtualRoom: Room | undefined;
 
+  public days: { [day: string]: boolean } = {
+    Mon: false,
```
Comment (ajaygandecha) [inline frontend/src/app/my-courses/course/office-hours/office-hours-editor/office-hours-editor.component.ts:60]: Making DoWs as an enum might help to prevent future issues using strings as keys.

File: frontend/src/app/my-courses/course/office-hours/office-hours.resolver.ts [hunk 2]
```diff
@@ -29,7 +29,8 @@ export const officeHoursResolver: ResolveFn<OfficeHours | undefined> = (
       start_time: new Date(),
       end_time: new Date(),
       course_site_id: +route.paramMap.get('course_site_id')!,
-      room_id: ''
+      room_id: '',
+      recurrence_id: -1
```
Comment (ajaygandecha) [inline frontend/src/app/my-courses/course/office-hours/office-hours.resolver.ts:33]: Should this be `null` instead?

File: backend/test/services/office_hours/office_hours_recurrence_test.py
Comment (ajaygandecha) [inline backend/test/services/office_hours/office_hours_recurrence_test.py:1]: Looking great! We are at 93% coverage with missing coverage for thrown `RecurringOfficeHourEventException`s on lines `124` and `141` of `backend/services/office_hours/office_hours_recurrence.py`.

**Review by ajaygandecha — CHANGES_REQUESTED** [review]
Great start @jadekeegan, amazing to see this come together! 🥳 

Added review comments for the backend. I am unable to test this functionality on the frontend despite resetting the database, trying to log in as Ina or Root, creating a new course site. This might be due to errors with the test data! When clicking "Create Event", I get the following error:

```
TypeError: undefined is not an object (evaluating 'this.myCoursesService.currentTerms()[0].end') — 
```
This happens even when trying to edit the term's end time to be past today. I can review the frontend once this gets fixed!

File: backend/entities/office_hours/office_hours_recurrence_entity.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+from sqlalchemy import Date, Integer, Boolean
+from datetime import date
+
+__authors__ = [
+    "Jade Keegan"
+]
+__copyright__ = "Copyright 2024"
+__license__ = "MIT"
+
+class OfficeHoursRecurrenceEntity(EntityBase):
```
[the remark this answers is not in the document]
Comment (ajaygandecha) [reply backend/entities/office_hours/office_hours_recurrence_entity.py:15]: (This also means a recommendation to rename entities and models from `OfficeHoursRecurrence` --> `OfficeHoursRecurrencePattern`!)

File: backend/services/office_hours/office_hours_recurrence.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        if recurrence.recur_saturday:
+            days_recur.append("saturday")
+
+        if recurrence.recur_sunday:
+            days_recur.append("sunday")
+
+        if len(days_recur) == 0:
+            raise RecurringOfficeHourEventException("No days to recur selected.")
+
+        # error out if recurrence end date is before 1st OH event
```
[the remark this answers is not in the document]
Comment (jadekeegan) [reply backend/services/office_hours/office_hours_recurrence.py:98]: Handled in the exception at the end!

**Review by ajaygandecha — APPROVED** [review]
Code LGTM, will bubble this up to @KrisJordan for the final approval!

**Comment by KrisJordan** [comment]
LGTM as well, will add the migration and get deployed for Spring. Great work @jadekeegan !

**Comment by KrisJordan** [comment]
@jadekeegan after code review I went to validate via click through and ran into a few minor concerns. Some of these can be punted on, but I think the first might be a blocker.

1. When attempting to save a recurring event, I'm currently getting an error that Python can't compare offset-naive and offset-aware datetimes here: https://github.com/unc-csxl/csxl.unc.edu/blob/task/CSXL-597/backend/services/office_hours/office_hours_recurrence.py#L99

Perhaps this is a Chrome widgets issue? The date is selected by a date selector and time through the date + time picker. Based on network tab, this appears to be the case:

Screenshot 2024-12-30 at 10 12 28 PM

2. When selecting recurring days, the checkbox makes dates scroll out of view: 
Screenshot 2024-12-30 at 10 14 35 PM

Perhaps we drop the check and just use shading to indicate?

3. Edge case: If you attempt to create a new event for a course after the semester has ended, you get a blank screen with no error message. Some kind of error message is probably appropriate here.

**Comment by KrisJordan** [comment]
I believe I've fixed the checkbox issue, but actually made a regression on the JavaScript time issue. Will revisit in the morning; at present (with West Coast local offset) I seem to be compounding a timezone offset issue between front end and backend.

Migration ready to go otherwise.

**Comment by ajaygandecha** [comment]
Wait @KrisJordan / @jadekeegan, curious of the status on this PR!

**Comment by ajaygandecha** [comment]
@KrisJordan and @jadekeegan, did another test of this feature on both Safari and Chrome for various different patterns and edge cases -- I am not getting any errors, things seem to be working.

I caught only one missing edge case - the frontend errored out in the case where the recurrence end date was after the first event's start time, but due to the pattern, no events could be created before the end date.

For example: 
- Today is Fri
- Attempt to create a recurrent event with pattern M Tu, but the recurrent end date is Sun
- This returned an empty of list events `[]` rather than erroring out, so I added a check in the backend to account for this.

Other than this, this still looks good to me. @KrisJordan, anything else you found that needs fixing?
